channels/urbdrc/client/urbdrc_main.c

```diff
@@ -59,9 +59,9 @@ static int func_hardware_id_format(IUDEVICE* pdev, char(*HardwareIds)[DEVICE_HAR
 	idProduct = (UINT16)pdev->query_device_descriptor(pdev, ID_PRODUCT);
 	bcdDevice = (UINT16)pdev->query_device_descriptor(pdev, BCD_DEVICE);
 	sprintf_s(str, sizeof(str), "USB\\VID_%04"PRIX16"&PID_%04"PRIX16"", idVendor, idProduct);
-	strcpy(HardwareIds[1], str);
+	strncpy(HardwareIds[1], str, DEVICE_HARDWARE_ID_SIZE);
 	sprintf_s(str, sizeof(str), "%s&REV_%04"PRIX16"", HardwareIds[1], bcdDevice);
-	strcpy(HardwareIds[0], str);
+	strncpy(HardwareIds[0], str, DEVICE_HARDWARE_ID_SIZE);
```

Comment:
Since both str and HardwareIds[0] are declared to be char[DEVICE_HARDWARE_ID_SIZE] and we ensure that str is null-terminated with sprintf_s, there's no reason to use strncpy here (and in similar situations).
At the very least, if you want to prevent problems if a change in the source made str longer, you should:
1- use sizeof(HardwareIds[0]) to make sure a similar change of the HardwarIds declaration is also automatically reported to the strncpy calls,
2- compare and check the sizes of HardwareIds[0] and str, because if str is longer, **strncpy WILL NOT NULL-TERMINATE the destination string !!!**
The later property of strncpy is the reason why I would in general advise against strncpy; this is a too low-level function for common use.

If you want both the check and ensure sturdiness in case of inconsistent changes in the sources, then you should use an abstract data type for string objects, with run-time checks implemented.

Reply:
@informatimago good point, will change to `sprintf_s`